src/core/cgroup.c

```diff
@@ -740,19 +746,27 @@ static void cgroup_context_apply(Unit *u, CGroupMask mask, ManagerState state) {
         if ((mask & CGROUP_MASK_MEMORY) && !is_root) {
                 if (cg_unified() > 0) {
                         uint64_t max = c->memory_max;
+                        uint64_t swap_max = c->memory_swap_max;
 
-                        if (cgroup_context_has_unified_memory_config(c))
+                        if (cgroup_context_has_unified_memory_config(c)) {
                                 max = c->memory_max;
-                        else {
+                                swap_max = c->memory_swap_max;
+                        } else {
                                 max = c->memory_limit;
 
                                 if (max != CGROUP_LIMIT_MAX)
                                         log_cgroup_compat(u, "Applying MemoryLimit %" PRIu64 " as MemoryMax", max);
+
+                                swap_max = c->memory_swap_limit;
+
```

Comment:
I'd followed the `max = c->memory_limit;` for `swap_max`.  But I'm agreeing to your opinion. Because the unified cgroup swap is operating different with legacy cgroup as menthioned at:
> The combined memory+swap accounting and limiting is replaced by real
control over swap space.
https://www.kernel.org/doc/Documentation/cgroup-v2.txt

So memory_limit and memory_swap_limit have to be used only for legacy cgroup. And memory_max and memory_swap_max have to be used only for unified cgroup.

I will rework this soon.

Thanks you for comments.